src/utilities/createRouterRoute.ts

```diff
@@ -1,6 +1,32 @@
 import { readonly } from 'vue'
-import { ResolvedRoute, RouterRoute } from '@/types'
+import { ResolvedRoute } from '@/types/resolved'
+import { RouteUpdate, RouteUpdateOptions, RouterUpdate } from '@/types/routerUpdate'
 
-export function createRouterRoute(route: ResolvedRoute): RouterRoute {
-  return readonly(route)
+export type RouterRoute<TRoute extends ResolvedRoute = ResolvedRoute> = Omit<ResolvedRoute, 'params'> & Readonly<{
+  params: { -readonly [P in keyof TRoute['params']]: TRoute['params'][P] },
+  update: RouteUpdate<TRoute>,
+}>
+
+export function createRouterRoute<TRoute extends ResolvedRoute>(route: TRoute, update: RouterUpdate): RouterRoute<TRoute> {
+  function routeUpdate(keyOrParams: PropertyKey | Partial<ResolvedRoute['params']>, valueOrOptions: any, maybeOptions?: RouteUpdateOptions): Promise<void> {
+    if (typeof keyOrParams === 'object') {
+      return update(route, keyOrParams, valueOrOptions)
+    }
+
+    const updatedParams = readonly({
+      ...route.params,
+      [keyOrParams]: valueOrOptions,
+    })
+    return update(route, updatedParams, maybeOptions)
```

Comment:
I'm guessing the types for update need the writable type so that here we don't have to first pass the params into readonly.

Reply:
yep, good catch